tools/qlog/src/lib.rs

```diff
@@ -258,6 +298,264 @@ impl Default for Qlog {
+            },
+
+            _ => return Err(Error::Done),
+        }
+
+        Ok(())
+    }
+
+    /// Finishes qlog streaming serialization.
+    ///
+    /// The JSON-serialized output has remaining close delimiters added.
+    /// After this is called, no more serialization will occur.
+    pub fn log_finish(&mut self) -> Result<()> {
+        if !self.log_started || self.log_finished {
+            return Err(Error::InvalidState);
+        }
+
+        self.writer.as_mut().write_all(b"]}]}")?;
+
+        self.log_finished = true;
+
+        self.writer.as_mut().flush()?;
+
+        Ok(())
+    }
+
+    /// Writes a JSON-serialized `EventField`s.
+    ///
+    /// Some qlog events can contain `QuicFrames`. If this is detected `true` is
+    /// returned and the streamer enters a frame-serialization mode that is only
+    /// concluded by `finish_frames()`. In this mode, attempts to log additional
+    /// events are ignored.
+    ///
+    /// If the event contains no array of `QuicFrames` return `false`.
+    pub fn write_event(&mut self, event: event::Event) -> Result<bool> {
+        if !self.log_started ||
+            self.log_finished ||
+            self.writing_event ||
+            self.writing_frames
+        {
+            return Err(Error::InvalidState);
+        }
+
+        let event_time = self.start_time.elapsed();
+
+        if let Some(t) = self.qlog.traces.get(0) {
+            let rel = match &t.configuration {
+                Some(conf) => match conf.time_units {
+                    Some(TimeUnits::Ms) => event_time.as_millis().to_string(),
+
+                    Some(TimeUnits::Us) => event_time.as_micros().to_string(),
+
+                    None => String::from(""),
+                },
+
+                None => String::from(""),
+            };
+
+            // TODO: this unfortunately removes any frames already existing in the
+            // event, which might be undesirable.
+            let (ev_data, contains_frames) =
+                match serde_json::to_string(&event.data) {
+                    Ok(mut ev_data_out) =>
+                        if event.data.contains_quic_frames() {
+                            if let Some(index) = ev_data_out.find("\"frames\":[")
+                            {
+                                ev_data_out
+                                    .truncate(index + "\"frames\":[".len());
+                            }
+
+                            self.first_frame = true;
+
+                            (ev_data_out, true)
+                        } else {
+                            (ev_data_out, false)
+                        },
+
+                    _ => return Err(Error::Done),
+                };
+
+            let maybe_comma = if self.first_event {
+                self.first_event = false;
+                ""
+            } else {
+                ","
+            };
+
+            let maybe_terminate = if contains_frames { "" } else { "]" };
```

Comment:
Related to the other comment about `Serialize`, if we manage to use `serialize_seq()` for this we would remove quite a bit of code here I think.

Reply:
Maybe I'm missing something but AFIACT to use `serialize_seq()` I would need to define manual serialization code and find some way to keep the `SerliazeSeq` around so that it can be invoked as frames and when frames need to be streamed. 

The autogenerated serialization code for `EventData` today is ~4k lines. And the qlog specification for this datatype is iikely to churn.